contracts/ito.sol

```diff
@@ -239,18 +327,31 @@ contract HappyTokenPool {
         require(withdraw_balance > 0, "None of this token left");
         uint256 expiration = unbox(pool.packed1, 232, 24) + base_timestamp;
         uint256 remaining_tokens = unbox(pool.packed2, 0, 128);
+        // only after expiration or the pool is empty
         require(expiration <= block.timestamp || remaining_tokens == 0, "Not expired yet");
         address token_address = pool.exchange_addrs[addr_i];
 
+        // ERC20
         if (token_address != DEFAULT_ADDRESS)
             transfer_token(token_address, address(this), msg.sender, withdraw_balance);
+        // ETH
         else
             msg.sender.transfer(withdraw_balance);
+        // clear the record
         pool.exchanged_tokens[addr_i] = 0;
```

Comment:
IDK could reset more variants to refund more gas?

```
 pool.ratios[addr_i*2] = 0;
 pool.ratios[addr_i*2+1] = 0;
pool.exchange_addrs[addr_i] = DEFAULT_ADDRESS;
```

Reply:
Users will need to call `destruct()` after all so these variables will be taken care of in that transaction.